src/python/WMComponent/WorkflowUpdater/WorkflowUpdaterPoller.py

```diff
@@ -26,6 +32,160 @@
+    :param blocks: list of blocks
+    :return: directionary
+    """
+    # rucio returns: [{"name": blockName, "replica": list(set(rses))}, ...]
+    rucioList = rucio.getReplicaInfoForBlocks(block=blocks)
+    bdict = {}
+    for item in rucioList:
+        bdict[item['name']] = item['replica']
+    return bdict
+
+
+def checkChanges(rdict, bdict):
+    """
+    Check if there is actually any pileup changes
+    :param rdict: dict of blocks locations from rucio
+    :param bdict: dict of blocks locations from pileupconf.json sandbox file
+    :return: boolean
+    """
+    # compare block names
+    if len(rdict) != len(bdict):
+        return True
+
+    # compare lcoations
+    for key in rdict.keys():
+        if sorted(rdict[key]) != sorted(bdict[key]):
+            return True
+    return False
+
+
+def updateBlockInfo(jdoc, rdict, bdict):
+    """
+    Discard given pileup names from JSON sandbox and return new dict.
+    NOTE: this function may consume lots of memory due to jdoc structure which
+    we need to parse and create new dict since we can't discard keys in place 
+    of jdoc internal dictionary, i.e.
+
+    jdoc structure is
+    {"<type: mc>": {"<block>": {"FileList" [], "NumberOfEvents":1, "PhEDExNodeNames": []},
+                    "<block>": {"FileList" [], "NumberOfEvents":1, "PhEDExNodeNames": []}, ...}
+    and we need to discard block names. They are located in internal dictionary
+    and in place pop-up of jdoc is not allowed since it will change size of internal dictionary.
+    Therefore, to overcome this obstacles we need to copy relevant pileupName key:values
+    into new dict structure and return new dict. Such operation will mostly double the
+    memory size.
+
+    :param jdoc: JSON sandbox dictionary
+    :param rdict: dict of blocks locations from rucio
+    :param bdict: dict of blocks locations from pileupconf.json sandbox file
+    :return: newly constructed dict
+    """
+    newDict = {}
+    for key, rec in jdoc.items():
+        newRec = {}
+        for blk in rec.keys():
+            # check if block present in the new container
+            if blk not in bdict.keys():
+                continue
+            loc = rec[blk]['PhEDExNodeNames']
+            rses = rdict.get(blk, bdict[blk])
+            if loc != rses:
+                rec[blk]['PhEDExNodeNames'] = rses
+            newRec[blk] = rec[blk]
+        newDict[key] = newRec
+    return newDict
+
+
+def writePileupJson(tfile, jdict, dest=None):
+    """
+    Write pileup JSON sandbox files back to file system
+    :param tfile: tar ball file name
+    :param jdict: JSON sandbox dictionary in form: {"path-to-json1": jdoc1, "path-to-json2": jdoc2}
+    :param dest: optional destination parameter to write final tar ball (use for unit tests)
+    :return: nothing
+    """
+    bname = os.path.basename(tfile)
+    dname = os.path.dirname(tfile)
+    ofile = f"{dname}/new-{bname}"
+    with tempfile.TemporaryDirectory() as tmpDir:
```

Comment:
I understand that. My question is whether we really have to add all of that logic inside a context manager(?) IMO, context manager is only needed for IO operations, when manipulating data in memory there is no need to have it inside the context manager block. Am I missing something?

Reply:
Yes, it still required since we extract tar ball into temp dir. If during this process and when we manipulate the data the exception will be thrown then temp dir will be left on a system and therefore
both its location and content will not be cleaned up after exception. This is IO leak. The context is much more elegant solution wrt to try/except and manual clean-up (it will be more code lines, more code, etc.).